test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,271 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @bug 8343791
+ * @summary verifies the socket is closed on `connect()` failures
+ * @library /test/lib
+ * @run junit/othervm --add-opens java.base/java.net=ALL-UNNAMED CloseOnFailureTest
+ */
+class CloseOnFailureTest {
+
+    private static final VarHandle SOCKET_IMPL_FACTORY_HANDLE = createSocketImplFactoryHandle();
+
+    private static VarHandle createSocketImplFactoryHandle() {
+        try {
+            Field field = Socket.class.getDeclaredField("factory");
+            field.setAccessible(true);
+            return MethodHandles
+                    .privateLookupIn(Socket.class, MethodHandles.lookup())
+                    .findStaticVarHandle(Socket.class, "factory", SocketImplFactory.class);
+        } catch (NoSuchFieldException | IllegalAccessException error) {
+            throw new RuntimeException(error);
+        }
+    }
+
+    private static void withSocketImplFactory(SocketImplFactory newFactory, ThrowingRunnable runnable) throws Throwable {
+        SocketImplFactory oldFactory = (SocketImplFactory) SOCKET_IMPL_FACTORY_HANDLE.getAndSet(newFactory);
+        try {
+            runnable.run();
+        } finally {
+            SOCKET_IMPL_FACTORY_HANDLE.set(oldFactory);
+        }
+    }
+
+    private static final class MockSocketImpl extends ThrowingSocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindError;
+
+        private final Exception connectError;
+
+        private MockSocketImpl(Exception bindError, Exception connectError) {
+            this.bindError = bindError;
+            this.connectError = connectError;
+        }
+
+        @Override
+        protected void create(boolean stream) {}
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindError);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectError);
+        }
+
+        private void throwIfPresent(Exception connectError) throws IOException {
+            if (connectError != null) {
+                switch (connectError) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown connectError type: " + connectError.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("ctor_should_close_on_failures")
+    @SuppressWarnings("resource")
+    void ctor_should_close_on_failures(TestCase testCase) throws Throwable {
```

Comment:
Is the style convention for  Java code  to use camel case for all classes, methods and variables ?

Reply:
Switched back to camel-case in e37ee840b9681aa017f8653df68c69232a6b6c83. (Snake-case convention for test method names is something I adopted several years ago from other projects, since it was helping with making the failure output more human-readable. But I'd be more than happy to stick to the established convention in the code base.)